src/coreclr/jit/lclvars.cpp

```diff
@@ -873,9 +874,89 @@ void Compiler::lvaInitUserArgs(InitVarDscInfo* varDscInfo, unsigned skipArgs, un
+        uint32_t  floatFlags          = STRUCT_NO_FLOAT_FIELD;
+        var_types argRegTypeInStruct1 = TYP_UNKNOWN;
+        var_types argRegTypeInStruct2 = TYP_UNKNOWN;
+
+        if ((strip(corInfoType) == CORINFO_TYPE_VALUECLASS) && (argSize <= MAX_PASS_MULTIREG_BYTES))
+        {
+            floatFlags = info.compCompHnd->getLoongArch64PassStructInRegisterFlags(typeHnd);
+        }
+
+        if ((floatFlags & STRUCT_HAS_FLOAT_FIELDS_MASK) != 0)
+        {
+            assert(varTypeIsStruct(argType));
+            int floatNum = 0;
+            if ((floatFlags & STRUCT_FLOAT_FIELD_ONLY_ONE) != 0)
+            {
+                assert(argSize <= 8);
+                assert(varDsc->lvExactSize <= argSize);
+
+                floatNum              = 1;
+                canPassArgInRegisters = varDscInfo->canEnreg(argRegTypeInStruct1, 1);
+
+                argRegTypeInStruct1 = (varDsc->lvExactSize == 8) ? TYP_DOUBLE : TYP_FLOAT;
+            }
+            else if ((floatFlags & STRUCT_FLOAT_FIELD_ONLY_TWO) != 0)
+            {
+                floatNum              = 2;
+                canPassArgInRegisters = varDscInfo->canEnreg(TYP_DOUBLE, 2);
+
+                argRegTypeInStruct1 = (floatFlags & STRUCT_FIRST_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
+                argRegTypeInStruct2 = (floatFlags & STRUCT_SECOND_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
+            }
+            else if ((floatFlags & STRUCT_FLOAT_FIELD_FIRST) != 0)
+            {
+                floatNum              = 1;
+                canPassArgInRegisters = varDscInfo->canEnreg(TYP_DOUBLE, 1);
+                canPassArgInRegisters = canPassArgInRegisters && varDscInfo->canEnreg(TYP_I_IMPL, 1);
+
+                argRegTypeInStruct1 = (floatFlags & STRUCT_FIRST_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
+                argRegTypeInStruct2 = (floatFlags & STRUCT_SECOND_FIELD_SIZE_IS8) ? TYP_LONG : TYP_INT;
+            }
+            else if ((floatFlags & STRUCT_FLOAT_FIELD_SECOND) != 0)
+            {
+                floatNum              = 1;
+                canPassArgInRegisters = varDscInfo->canEnreg(TYP_DOUBLE, 1);
+                canPassArgInRegisters = canPassArgInRegisters && varDscInfo->canEnreg(TYP_I_IMPL, 1);
+
+                argRegTypeInStruct1 = (floatFlags & STRUCT_FIRST_FIELD_SIZE_IS8) ? TYP_LONG : TYP_INT;
+                argRegTypeInStruct2 = (floatFlags & STRUCT_SECOND_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
+            }
+
+            assert((floatNum == 1) || (floatNum == 2));
+
+            if (!canPassArgInRegisters)
+            {
+                // On LoongArch64, if there aren't any remaining floating-point registers to pass the argument,
+                // integer registers (if any) are used instead.
+                canPassArgInRegisters = varDscInfo->canEnreg(argType, cSlotsToEnregister);
+
+                argRegTypeInStruct1 = TYP_UNKNOWN;
+                argRegTypeInStruct2 = TYP_UNKNOWN;
```

Comment:
I wonder if the split case is handled correctly. Consider `struct { double, long }`, and we have exactly one integer register available.

First, we will fall into this `if`, and ask `canEnreg(TYP_STRUCT, 2);` aka `canEnreg(TYP_I_IMPL, 2)`, that will return `false` and we will pass the whole thing on the stack, but is that correct?

```
GRLEN < WOA ≤ 2 × GRLEN

- The structure has one double member and only one fixed-point member.

If one FAR and one GAR are available, the floating-point member of the structure is passed in the FAR,
and the integer member of the structure is passed in the GAR; If no floating-point registers but two
GARs are available, it’s passed in the two GARs; If only one GAR is available, the low-order bits are
in the GAR and the high-order bits are on the stack; And it’s passed on the stack if no GAR is available.
```

Reply:
Yes, you are right, I have to amend this.

Thanks very much.

I will update it now.

Reply:
Hi, @SingleAccretion 
I had updated within the patch https://github.com/dotnet/runtime/pull/62843/commits/5c10577b66880682e742f7765a976b1463075c0a

Thanks very much.

Reply:
@shushanhf Depending on how complete your bringup of the runtime is you may be interested in running the [ABI stress tests](https://github.com/dotnet/runtime/tree/main/src/tests/JIT/Stress/ABI). You can invoke it with various command line arguments (see the .csproj files) and it will test invoking functions with many permutations of arguments and verify that arguments are being passed through correctly. It should quickly find issues like this one.

Reply:
> @shushanhf Depending on how complete your bringup of the runtime is you may be interested in running the [ABI stress tests](https://github.com/dotnet/runtime/tree/main/src/tests/JIT/Stress/ABI). You can invoke it with various command line arguments (see the .csproj files) and it will test invoking functions with many permutations of arguments and verify that arguments are being passed through correctly. It should quickly find issues like this one.

Thanks very much.